conans/client/graph/graph_binaries.py

```diff
@@ -206,12 +206,10 @@ def _process_node(self, node, build_mode, remotes, update):
                 break
 
         if node.conanfile.upload_policy == "skip":
+            # Download/update shouldn't be checked in the servers if this is "skip-upload"
```

Comment:
I don't understand why **upload**_policy should affect **download**/update in any way. Why `if node.conanfile.upload_policy == "skip":` is needed here and cannot be removed completely?

Reply:
Because if a binary cannot be uploaded, it doesn't make sense that Conan would be looking for binaries in remotes, so Conan should avoid doing that. Some users were requesting that Conan should be able to work offline, without trying to reach the servers at all for upload-policy=skip, and it makes sense. So download of packages must check it.

Reply:
> Because if a binary cannot be uploaded, it doesn't make sense that Conan would be looking for binaries in remotes, so Conan should avoid doing that.

Just to save some time? With `upload_policy=skip`, the binaries simply can't end up on the server. I don't think this kludge is worth it, but of course, I may not understand something.

> Some users were requesting that Conan should be able to work offline, without trying to reach the servers at all for upload-policy=skip, and it makes sense. So download of packages must check it.

The recipe itself still has to be downloadable/updatable,  `upload_policy` just affects binaries. So it seems to help the offline work not much, Conan servers still need to be accessible. And if `upload_policy=skip` were triggering the fully offline mode, one would need **all** his/her recipes to have this attribute? I don't think it's a realistic scenario.

Reply:
> Just to save some time? With upload_policy=skip, the binaries simply can't end up on the server. I don't think this kludge is worth it, but of course, I may not understand something.

Not for performance, but for "offline" mode. Conan should be able to work offline if it has everything it needs in the cache, and this is the case if ``upload_policy=skip``.

> The recipe itself still has to be downloadable/updatable, upload_policy just affects binaries. So it seems to help the offline work not much, Conan servers still need to be accessible. And if upload_policy=skip were triggering the fully offline mode, one would need all his/her recipes to have this attribute? I don't think it's a realistic scenario.

But the recipe can already be in the cache without the binary. This can happen in several scenarios, for example with a ``conan export``. The problem is that if every ``conan install`` doesn't find binaries in the cache it will go and check in the server, and then fail if there is no network or something like that. The current implementation makes this case work smoothly as users expect, not checking in the servers if not necessary.

Reply:
> The current implementation makes this case work smoothly as users expect, not checking in the servers if not necessary.

This use of `upload_policy=skip` is not documented in any way. Should it? I wouldn't say this behavior is quite obvious.  And you write `# Download/update shouldn't be checked in the servers if this is "skip-upload"`. Why "update"? It only applies to recipes, not binaries, and is not prevented by `upload_policy=skip`.

Reply:
It is assumed to be the default and expected behavior, not hitting the servers when not necessary, this is why it is not documented that explicitly. But we can definitely improve the docs in to clarify this https://docs.conan.io/2/reference/conanfile/attributes.html#upload-policy

Reply:
Tracking in https://github.com/conan-io/docs/issues/3507